core/client/db/src/storage_cache.rs

```diff
@@ -654,6 +666,62 @@ mod tests {
 		assert!(s.storage(&key).unwrap().is_none());
 	}
 
+	#[test]
+	fn simple_fork() {
+		//init_log();
+		let root_parent = H256::random();
+		let key = H256::random()[..].to_vec();
+		let h1 = H256::random();
+		let h2a = H256::random();
+		let h2b = H256::random();
+
+		let shared = new_shared_cache::<Block, Blake2Hasher>(256*1024, (0,1));
+
+		let mut s = CachingState::new(InMemory::<Blake2Hasher>::default(), shared.clone(), Some(root_parent.clone()));
+		s.cache.sync_cache(&[], &[], vec![(key.clone(), Some(vec![2]))], vec![], Some(h1.clone()), Some(1), || true);
+
+		let mut s = CachingState::new(InMemory::<Blake2Hasher>::default(), shared.clone(), Some(h1.clone()));
+		s.cache.sync_cache(&[], &[], vec![(key.clone(), Some(vec![2]))], vec![], Some(h2a.clone()), Some(2), || true);
+
+		let mut s = CachingState::new(InMemory::<Blake2Hasher>::default(), shared.clone(), Some(h1.clone()));
+		s.cache.sync_cache(&[], &[], vec![(key.clone(), Some(vec![3]))], vec![], Some(h2b.clone()), Some(2), || false);
+
+		let s = CachingState::new(InMemory::<Blake2Hasher>::default(), shared.clone(), Some(h2a.clone()));
+		assert_eq!(s.storage(&key).unwrap().unwrap(), vec![2]);
+	}
+
+	#[test]
+	fn double_fork() {
+		//init_log();
+		let root_parent = H256::random();
+		let key = H256::random()[..].to_vec();
+		let h1 = H256::random();
```

Comment:
This is a nit that you can ignore but I would def. consider if I can hide this process of generating `N` random `h256::random()`s somewhere. In the easiest case, if you are fine with using them behind an index (as opposed to h1, h1a etc.) you can do something along the lines of `(0..N).map(H256::random).collect()` or sth similar.

Reply:
it will be hard to semantically distinguish forks and so on but thanks!